### PR TITLE
Feat/improve yaml extraction

### PR DESCRIPTION
**Pull request by bhancockio**

### CODE REVIEW

**Comment by joaomdmoura** [comment]
Disclaimer: This review was made by a crew of AI Agents.

# Code Review Comment for PR #2428

## Overview
This pull request introduces notable enhancements to the event handling system and string interpolation implementation within the `crewAI` repository. The changes significantly improve both flexibility and robustness, but I have identified several areas for improvement.

### 1. Event Bus Enhancements (`crewai_event_bus.py`)

**Positive Aspects:**
- The introduction of wildcard event handling via `isinstance()` enhances flexibility in event management.
- Unused variables have been pruned, resulting in cleaner code and a more maintainable structure.
- Application of the singleton pattern is cleanly executed.

**Areas for Improvement:**
- **Testing Handler Cleanup**: The removal of the `clear_handlers()` method may complicate testing. I recommend retaining it as an internal method:
  ```python
  def _clear_handlers(self) -> None:
      """Internal method to clear handlers for testing purposes."""
      self._handlers.clear()
  ```

- **Performance Optimization**: The current handler iteration could be improved by maintaining a class hierarchy map to reduce the performance impact when emitting events. An example to consider:
  ```python
  class CrewAIEventsBus:
      def __init__(self):
          self._handlers = defaultdict(list)
          self._class_hierarchy = defaultdict(set)
      
      def register_handler(self, event_type: Type[EventTypes], handler: Callable):
          self._handlers[event_type].append(handler)
          for base in event_type.__mro__[1:]:
              if issubclass(base, CrewEvent):
                  self._class_hierarchy[base].add(event_type)
      
      def emit(self, source: Any, event: CrewEvent) -> None:
          event_type = type(event)
          for handler in self._handlers[event_type]:
              handler(source, event)
          for parent_type in event_type.__mro__[1:]:
              if parent_type in self._handlers:
                  for handler in self._handlers[parent_type]:
                      handler(source, event)
  ```

### 2. String Interpolation Improvements (`formatter.py`)

**Positive Aspects:**
- The validation of input types is robust, ensuring that only correctly formatted data is processed.
- The decision to separate interpolation logic into `formatter.py` greatly enhances code organization.
- Edge cases are well-handled, increasing the reliability of the functionality.

**Areas for Improvement:**
- **Error Message Clarity**: Consider providing more detailed error messages to help diagnose issues:
  ```python
  def interpolate_only(input_string: Optional[str], inputs: Dict[str, Any]) -> str:
      if not input_string:
          return ""
      if not inputs:
          raise ValueError(
              "Inputs dictionary is empty. Required variables: "
              f"{re.findall(r'\{([A-Za-z_][A-Za-z0-9_]*)\}', input_string)}"
          )
  ```

- **Performance Optimization**: Use a cached compiled regex pattern for better performance:
  ```python
  _VARIABLE_PATTERN = re.compile(r"\{([A-Za-z_][A-Za-z0-9_]*)\}")
  ```

### 3. Test Coverage Enhancements (`test_crewai_event_bus.py`, `test_formatter.py`)

**Positive Aspects:**
- Adequate test coverage ensures that changes are adequately validated.
- Test organization is clear, which aids in maintainability.
- A good mix of unit tests is included, enhancing the overall quality of the code.

**Areas for Improvement:**
- **Missing Edge Cases**: Additional tests for concurrent event emissions would be beneficial. Here’s an example:
  ```python
  def test_concurrent_event_emission():
      mock_handler = Mock()
      events = [TestEvent(type="test_event") for _ in range(100)]
    
      @crewai_event_bus.on(TestEvent)
      def handler(source, event):
          mock_handler(source, event)
    
      with ThreadPoolExecutor(max_workers=10) as executor:
          futures = [executor.submit(crewai_event_bus.emit, "source", event) for event in events]
          wait(futures)
    
      assert mock_handler.call_count == 100
  ```
  
- **Test Isolation**: Make sure to implement setup and teardown routines for test isolation:
  ```python
  @pytest.fixture(autouse=True)
  def setup_event_bus():
      original_handlers = crewai_event_bus._handlers.copy()
      yield
      crewai_event_bus._handlers = original_handlers
  ```

## General Recommendations
1. Add type hints consistently across all new code for improved readability.
2. Introduce performance benchmarks for the event system under load.
3. Implement logging to assist in debugging complex event handling scenarios.
4. Consider introducing event prioritization to improve event processing.
5. Document the newly introduced wildcard event handling feature comprehensively.

## Security Considerations
1. Validate event payload sizes to prevent potential memory overflow issues.
2. Consider adding rate limiting for event emissions to mitigate DoS attacks.
3. Ensure input sanitization for strings being interpolated to prevent injection vulnerabilities.

The changes made in this PR are solid and represent a significant step forward for the project, especially regarding event handling flexibility and string interpolation robustness. Addressing the suggested improvements would further ensure maintainability and performance. 

---

Please review the suggested modifications and let me know if you need any further clarification or additional input regarding the changes made in this PR. Thank you!

File: src/crewai/utilities/string_utils.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import re
+from typing import Any, Dict, List, Optional, Union
+
+
+def interpolate_only(
```
Comment (lorenzejay) [inline src/crewai/utilities/string_utils.py:5]: nice job decoupling this from task

File: src/crewai/agents/agent_builder/base_agent.py [hunk 2]
```diff
@@ -333,9 +334,15 @@ def interpolate_inputs(self, inputs: Dict[str, Any]) -> None:
...
-            self.backstory = self._original_backstory.format(**inputs)
+            self.role = interpolate_only(
+                input_string=self._original_role, inputs=inputs
+            )
+            self.goal = interpolate_only(
+                input_string=self._original_goal, inputs=inputs
+            )
+            self.backstory = interpolate_only(
+                input_string=self._original_backstory, inputs=inputs
+            )
```
Comment (lorenzejay) [inline src/crewai/agents/agent_builder/base_agent.py:345]: nice!

**Review by lorenzejay — APPROVED** [review]
